amazon/dynamodb/client/src/main/kotlin/org/http4k/connect/amazon/dynamodb/mapper/DynamoDbQueryBuilder.kt

```diff
@@ -0,0 +1,286 @@
+        infix fun FilterExpression?.or(other: FilterExpression?): FilterExpression? = when {
+            this == null -> other
+            other == null -> this
+            else -> FilterExpression(
+                "($expression OR ${other.expression})",
+                attributeNames + other.attributeNames,
+                attributeValues + other.attributeValues
+            )
+        }
+
+        fun not(expr: FilterExpression?): FilterExpression? = expr?.let {
+            FilterExpression(
+                "(NOT ${it.expression})",
+                it.attributeNames,
+                it.attributeValues
+            )
+        }
+    }
+
+    private var attributeNameCount = 0
+
+    // generate consecutive names "a", "b", ... to be used as expression attribute name
+    private fun nextAttributeName(): String {
+        val base = 'z' - 'a' + 1
+        val name = StringBuilder()
+        var currentCount = attributeNameCount
+        do {
+            val remainder = currentCount % base
+            currentCount /= base
+            name.insert(0, 'a' + remainder)
+        } while (currentCount > 0)
+
+        attributeNameCount += 1
+        return name.toString()
+    }
+
+    private var keyCondition: KeyCondition? = null
+    private var filterExpression: FilterExpression? = null
+
+    fun keyCondition(block: KeyConditionBuilder.() -> CombinedKeyCondition) {
+        keyCondition = block(KeyConditionBuilder())
+    }
+
+    fun filterExpression(block: FilterExpressionBuilder.() -> FilterExpression?) {
+        filterExpression = block(FilterExpressionBuilder())
+    }
+
+    var scanIndexForward: Boolean = true
+    var pageSize: Int? = null
+    var consistentRead: Boolean? = null
+
+    private fun <K, V> union(map1: Map<K, V>?, map2: Map<K, V>?) = when {
+        map1 == null -> map2
+        map2 == null -> map1
+        else -> map1 + map2
+    }
+
+    internal fun build() = DynamoDbQuery(
+        keyConditionExpression = keyCondition?.expression,
+        filterExpression = filterExpression?.expression,
+        expressionAttributeNames = union(keyCondition?.attributeNames, filterExpression?.attributeNames),
+        expressionAttributeValues = union(keyCondition?.attributeValues, filterExpression?.attributeValues),
+        scanIndexForward = scanIndexForward,
+        pageSize = pageSize,
+        consistentRead = consistentRead
+    )
+}
+
+fun <Document : Any, HashKey : Any, SortKey : Any> DynamoDbIndexMapper<Document, HashKey, SortKey>.query(block: DynamoDbQueryBuilder.() -> Unit): Sequence<Document> {
+    val query = DynamoDbQueryBuilder().apply(block).build()
+    return query(
+        KeyConditionExpression = query.keyConditionExpression,
+        FilterExpression = query.filterExpression,
+        ExpressionAttributeNames = query.expressionAttributeNames,
+        ExpressionAttributeValues = query.expressionAttributeValues,
+        ScanIndexForward = query.scanIndexForward,
+        PageSize = query.pageSize,
```

Comment:
I wonder: is this intentional that this parameter is here called `PageSize` and in `queryPage` it is `Limit`?

Reply:
It's intentional.  The SDK call calls it `Limit`, but I feel like `PageSize` is less ambiguous when dealing with pages. However, whether it was correct to depart from the SDK naming conventions is uncertain 🤷 .